lib/AST/ASTImporter.cpp

```diff
@@ -2598,14 +2616,30 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   if (ImportTemplateInformation(D, ToFunction))
     return nullptr;
 
+  bool IsFriend = D->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend);
+
   // Add this function to the lexical context.
   // NOTE: If the function is templated declaration, it should be not added into
   // LexicalDC. But described template is imported during import of
   // FunctionTemplateDecl (it happens later). So, we use source declaration
   // to determine if we should add the result function.
-  if (!D->getDescribedFunctionTemplate())
+  if (!D->getDescribedFunctionTemplate() && !IsFriend)
     LexicalDC->addDeclInternal(ToFunction);
 
+  // Friend declarations lexical context is the befriending class, but the
+  // semantic context is the enclosing scope of the befriending class.
+  // We want the friend functions to be found in the semantic context by lookup.
+  // FIXME should we handle this genrically in VisitFriendDecl?
+  // In Other cases when LexicalDC != DC we don't want it to be added,
+  // e.g out-of-class definitions like void B::f() {} .
+  if (LexicalDC != DC && IsFriend) {
+    DC->makeDeclVisibleInContext(ToFunction);
+  }
+
+  // Import the rest of the chain. I.e. import all subsequent declarations.
+  for (++RedeclIt; RedeclIt != Redecls.end(); ++RedeclIt)
+    Importer.Import(*RedeclIt);
+
   if (auto *FromCXXMethod = dyn_cast<CXXMethodDecl>(D))
     ImportOverrides(cast<CXXMethodDecl>(ToFunction), FromCXXMethod);
 
```

Comment:
This way of importing ensures that the order of functions is kept as it was. But it results in a lot of recursive calls to import: Every import tries to import everything before and after the current function. When starting with importing a function in the middle of the chain, importing the first function in the chain will import already every other too (including the original one). Probably it is sufficient to import only the immediate previous or immediate next item in the chain (or both) since the importing continues then recursively.

Reply:
You are right, importing only the previous and the next decl would decrease the number of recursive calls in case of a long chain. However, I think, most of the time the chain is not longer than 2, usually we have a prototype and then a definition in the "from" context.